File: typescript/src/webhooks/retry.ts
import type { Logger } from "../types/index.js";
import { noopLogger, delay, calculateBackoff } from "../utils/index.js";

export interface DeliveryRecord {
  event: string;
  payload: unknown;
  attempts: number;
  lastError?: string;
}

const DEFAULT_MAX_RETRIES = 3;

export class WebhookRetryQueue {
  private queue: DeliveryRecord[] = [];
  private processing = false;
  private readonly logger: Logger;
  private readonly maxRetries: number;
  private readonly deadLetterQueue: DeliveryRecord[] = [];

  constructor(logger: Logger = noopLogger, maxRetries = DEFAULT_MAX_RETRIES) {
    this.logger = logger;
    this.maxRetries = maxRetries;
  }

  enqueue(event: string, payload: unknown): void {
    this.queue.push({ event, payload, attempts: 0 });
    this.logger.warn("Webhook enqueued for retry", { event });
    if (!this.processing) {
      this.processing = true;
      this.processQueue();
    }
  }

  private async processQueue(): Promise<void> {
    while (this.queue.length > 0) {
      const record = this.queue.shift()!;
      record.attempts++;

      try {
        await import("./index.js");
        this.logger.info("Retrying webhook delivery", {
          event: record.event,
          attempt: record.attempts,
        });
      } catch (err) {
        record.lastError = String(err);
        if (record.attempts < this.maxRetries) {
          const backoff = calculateBackoff(
            record.attempts - 1,
            1000,
            30000,
          );
          this.logger.warn("Webhook retry failed, re-enqueuing", {
            event: record.event,
            attempt: record.attempts,
            backoffMs: backoff,
          });
          this.queue.push(record);
          await delay(backoff);
        } else {
          this.logger.error("Webhook delivery failed, moving to DLQ", {
            event: record.event,
            attempts: record.attempts,
          });
          this.deadLetterQueue.push(record);
        }
      }
    }
    this.processing = false;
  }

  getDeadLetterQueue(): readonly DeliveryRecord[] {
    return this.deadLetterQueue;
  }
}

File: go/webhooks/retry.go
package webhooks

import (
	"math"
	"sync"
	"time"

	"github.com/yourdudeken/mpesa-sdk/go/types"
)

type DeliveryRecord struct {
	Event     string
	Payload   interface{}
	Attempts  int
	LastError string
}

type RetryQueue struct {
	mu              sync.Mutex
	queue           []*DeliveryRecord
	deadLetterQueue []*DeliveryRecord
	processing      bool
	logger          types.Logger
	maxRetries      int
}

func NewRetryQueue(logger types.Logger, maxRetries int) *RetryQueue {
	if logger == nil {
		logger = types.NewNoopLogger()
	}
	if maxRetries == 0 {
		maxRetries = 3
	}
	return &RetryQueue{
		logger:     logger,
		maxRetries: maxRetries,
	}
}

func (rq *RetryQueue) Enqueue(event string, payload interface{}) {
	rq.mu.Lock()
	rq.queue = append(rq.queue, &DeliveryRecord{Event: event, Payload: payload})
	rq.logger.Warn("Webhook enqueued for retry", "event", event)
	if !rq.processing {
		rq.processing = true
		go rq.processQueue()
	}
	rq.mu.Unlock()
}

func (rq *RetryQueue) processQueue() {
	for {
		rq.mu.Lock()
		if len(rq.queue) == 0 {
			rq.processing = false
			rq.mu.Unlock()
			return
		}
		record := rq.queue[0]
		rq.queue = rq.queue[1:]
		rq.mu.Unlock()

		record.Attempts++
		rq.logger.Info("Retrying webhook delivery",
			"event", record.Event,
			"attempt", record.Attempts,
		)

		if record.Attempts >= rq.maxRetries {
			rq.logger.Error("Webhook delivery failed, moving to DLQ",
				"event", record.Event,
				"attempts", record.Attempts,
			)
			rq.mu.Lock()
			rq.deadLetterQueue = append(rq.deadLetterQueue, record)
			rq.mu.Unlock()
		} else {
			backoff := time.Duration(math.Min(
				float64(1000)*math.Pow(2, float64(record.Attempts-1)),
				30000,
			)) * time.Millisecond
			rq.logger.Warn("Webhook retry failed, re-enqueuing",
				"event", record.Event,
				"attempt", record.Attempts,
				"backoff_ms", backoff.Milliseconds(),
			)
			time.Sleep(backoff)
			rq.mu.Lock()
			rq.queue = append(rq.queue, record)
			rq.mu.Unlock()
		}
	}
}

func (rq *RetryQueue) GetDeadLetterQueue() []*DeliveryRecord {
	rq.mu.Lock()
	defer rq.mu.Unlock()
	result := make([]*DeliveryRecord, len(rq.deadLetterQueue))
	copy(result, rq.deadLetterQueue)
	return result
}
